docs/specifications/tests/DNSSEC-TP/dnssec16.md

```diff
@@ -0,0 +1,323 @@
+          the name server IP address and key tag to 
+          *No Match CDS With DNSKEY* set.
+       3. Else, if there is no RRSIG for the DNSKEY RRset created by
+          the DNSKEY record that the CDS record points at then add the
+          name server IP address and key tag of CDS record to the
+          *DNSKEY Not Signed By CDS* set.
+    6. If there are no RRSIG records for the CDS RRset, then add the
+       name server IP address to the *CDS Not Signed* set.
+    7. Else, for each RRSIG (CDS) do:
+       1. If the key tag of the RRSIG does not match any DNSKEY then
+          add the name server IP address and key tag to the
+          *CDS Signed By Unknown DNSKEY* set.
+       2. Else, if the RRSIG cannot be validated by the DNSKEY it
+          refers to by key tag, then add the name server IP and RRSIG
+          key tag to the *CDS Invalid RRSIG* set.
+    8. Go to next name server IP address.
+
+9.  For each name server IP in the *CDNSKEY RRsets* set do:
+
+    1. Get the CDNSKEY and its RRSIG records, if any.
+    2. If any CDNSKEY record is a "delete" CDNSKEY, then do:
+       1. If there is more than a single CDNSKEY record then add the
+          name server IP to the *Mixed Delete CDNSKEY* set.
+       2. Else, add the name server IP address to the *Delete CDNSKEY*
+          set.
+       2. Go to next name server IP.
+    3. Get the DNSKEY and its RRSIG records, if any, from the
+       *DNSKEY RRsets* for the same name server IP.
+    4. If there are no DNSKEY records, then do:
+       1. Add name server IP address to the *No DNSKEY RRset* set
+          (duplicates not possible).
+       2. Go to next name server IP.
+    5. Repeat the following steps for each CDNSKEY record:
+       1. Compare the CDNSKEY record with the DNSKEY records.
+       2. If the CDNSKEY record does not match any DNSKEY record then
+          add the name server IP address and key tag to 
+          *No Match CDNSKEY With DNSKEY* set.
+       3. Else, if there is no RRSIG for the DNSKEY RRset created by
+          the DNSKEY record matching the CDNSKEY record then add the
+          name server IP address and key tag of CDNSKEY record to the
+          *DNSKEY Not Signed By CDNSKEY* set.
+    6. If there are no RRSIG records for the CDNSKEY RRset, then add
+       the name server IP address to the *CDNSKEY Not Signed* set.
+    7. Else, for each RRSIG (CDNSKEY) do:
+       1. If the key tag of the RRSIG does not match any DNSKEY then
+          add the name server IP address and key tag to the
+          *CDNSKEY Signed By Unknown DNSKEY* set.
+       2. Else, if the RRSIG cannot be validated by the DNSKEY it
+          refers to by key tag, then add the name server IP and RRSIG
+          key tag to the *CDNSKEY Invalid RRSIG* set.
+    8. Go to next name server IP address.
+
+10. If the *No DNSKEY RRset* set is non-empty, then output
+    *[DS16_CDS_CDNSKEY_WITHOUT_DNSKEY]* with all name server IP addresses
+    in the set.
```

Comment:
Positive statement "if all DNSKEY RRsets are empty ....". Much easier to understand

Reply:
The *No DNSKEY RRset* set is defined as a set of IP addresses.